lib/temporal/workflow/context.rb

```diff
@@ -215,14 +215,53 @@ def wait_for_all(*futures)
         return
       end
 
-      def wait_for(future)
+      # Block workflow progress until any future is finished or any unblock_condition
+      # block evaluates to true.
+      def wait_for(*futures, &unblock_condition)
+        if futures.empty? && unblock_condition.nil?
+          raise 'You must pass either a future or an unblock condition block to wait_for'
+        end
+
         fiber = Fiber.current
+        should_yield = false
+        blocked = true
+
+        if futures.any?
+          if futures.all?(&:finished?)
+            blocked = false
+          else
+            should_yield = true
+            futures.each do |future|
+              dispatcher.register_handler(future.target, Dispatcher::WILDCARD) do
```

Comment:
I wonder if we actually need to separate between the futures and unblock condition cases here? There are probably arguments to both approaches:

1. Only using a single wildcard handler for both cases means that it might get called more than needed, but it simplifies the code (only need to register one handler)
2. Using future-specific and a wildcard handlers is a bit of an optimisation for when only futures are passed, but is actually less optimal when both are passed

I wonder what do you think about this one?

Reply:
I believe this code path is used when invoking activities synchronously or sleeps, which are almost certainly the most common cases. I could register separate handlers here, rather than handling the separate cases inside the dispatcher. I'll give this alternative a try to see how it looks once implemented, especially with the unregistration in place.

Reply:
Sounds good 👍 But also happy to rearrange things in later PRs to keep this one contained and ship the feature